pkg/datastore/filedb/filedb.go

```diff
@@ -61,11 +62,80 @@ func (f *FileDB) fetch(ctx context.Context, path string) ([]byte, error) {
 }
 
 func (f *FileDB) Find(ctx context.Context, col datastore.Collection, opts datastore.ListOptions) (datastore.Iterator, error) {
-	_, ok := col.(datastore.ShardStorable)
+	scol, ok := col.(datastore.ShardStorable)
 	if !ok {
 		return nil, datastore.ErrUnsupported
 	}
-	return nil, datastore.ErrUnimplemented
+
+	var (
+		kind   = col.Kind()
+		shards = scol.ListInUsedShards()
+		// Map of objects values with the first key is the object id.
+		objects map[string][][]byte
+	)
+
+	// Fetch the first part of all objects under a specified "directory".
+	for _, shard := range shards {
+		dpath := makeHotStorageDirPath(kind, shard)
+		parts, err := f.backend.List(ctx, dpath)
+		if err != nil {
+			f.logger.Error("failed to find entities",
+				zap.String("kind", kind),
+				zap.Error(err),
+			)
+			return nil, err
+		}
+
+		if objects == nil {
+			objects = make(map[string][][]byte, len(parts))
+		}
+		for _, obj := range parts {
+			id := filepath.Base(obj.Path)
+
+			data, err := f.fetch(ctx, obj.Path)
```

Comment:
Oops. I was thinking that the List function also returns data of file objects so that we can have their contents without requiring any extra requests.
This way a lot of requests `number_of_shards * (1 + number_of_objects)` will be required in a short time and I think it will not be realistic.

Do we have any better idea for that problem?

Reply:
I feel you, the (n+1) problem, right. Tbh, I was thinking as you at first but when I read the place we use this filestore List interface (ref: [planpreview cleaner](https://github.com/pipe-cd/pipecd/blob/master/pkg/app/ops/planpreviewoutputcleaner/cleaner.go#L71)) I get the same surprise as you have now 😄 

I tried to work around with the "list" support API from our supporting filestore (gcs, s3 and minio), and looks like they only support the get object by key as the only way to fetch the raw data of an object, all list API seems only returns the attributes and meta which we can use to fetch the content. I will investigate more about that, but in the worst case, I think we have 2 points to rely on which are:
- the number of the object in hot storage of each kind is expected to be all enough, we may need to do some kind of middleware fetching objects parts in parallel here if necessary
- cache storage in API layer will reduce the number of times we have to list this in a direct way.

Wdyt?

Reply:
@nghialv thank you so much for your comment 🙏  I added logic to check whether the raw data of the object part is updated or not based on the etag value returned from `List` objects request. For now, we will only fetch for the object part stored under the given path in case there is no version of it found in `filedb.cache`.  PTAL when you have time 🙌

Reply:
Thank you. Let me take a look.